drivers/net/can/spi/mcp25xxfd/mcp25xxfd-core.c

```diff
@@ -2447,6 +2462,124 @@ mcp25xxfd_register_quirks(struct mcp25xxfd_priv *priv)
+	u32 osc, osc_reference, osc_mask;
+	int err;
+	u8 mode;
+
+	/* increase the usage count, and disallow suspend */
+	pm_runtime_get_sync(ndev->dev.parent);
+
+	/* Read MCP25XXFD_REG_OSC as the first thing if we are
+	 * possibly in sleep mode. If we read anything else first,
+	 * we run the risk of not hitting the workaround in
+	 * mcp25xxfd-regmap.c
+	 */
+	err = regmap_read(priv->map_reg, MCP25XXFD_REG_OSC, &osc);
+	if (err)
+		return err;
+
+	if (osc & MCP25XXFD_REG_OSC_OSCDIS) {
+		osc = FIELD_PREP(MCP25XXFD_REG_OSC_CLKODIV_MASK, MCP25XXFD_REG_OSC_CLKODIV_10);
+		err = regmap_write(priv->map_reg, MCP25XXFD_REG_OSC, osc);
+		if (err)
+			return err;
+	}
+
+	osc_reference = MCP25XXFD_REG_OSC_OSCRDY;
+	osc_mask = MCP25XXFD_REG_OSC_OSCRDY | MCP25XXFD_REG_OSC_PLLRDY;
+
+	/* Wait for "Oscillator Ready" bit */
+	err = regmap_read_poll_timeout(priv->map_reg, MCP25XXFD_REG_OSC, osc,
+					(osc & osc_mask) == osc_reference,
+					10000, 1000000);
+	if (err)
+		return err;
+
+	err = regmap_update_bits(priv->map_reg, MCP25XXFD_REG_OSC,
+			MCP25XXFD_REG_OSC_CLKODIV_MASK, priv->clkodiv);
+	if (err)
+		return err;
+
+	err = mcp25xxfd_chip_set_mode(priv, MCP25XXFD_REG_CON_MODE_CONFIG);
+	if (err)
+		return err;
+
+	return 0;
+}
+
+static void mcp25xxfd_oscout_unprepare(struct clk_hw *hw)
+{
+	struct mcp25xxfd_priv *priv =
+		container_of(hw, struct mcp25xxfd_priv, oscout);
+	struct net_device *ndev = priv->ndev;
+	u8 mode = 0;
+
+	/* if we have no more clock users, and the CAN is not active,
+	 * set the mode to SLEEP.
+	 */
+	mcp25xxfd_chip_get_mode(priv, &mode);
+	if (mode == MCP25XXFD_REG_CON_MODE_CONFIG) {
+		mcp25xxfd_chip_set_mode(priv, MCP25XXFD_REG_CON_MODE_SLEEP);
+	}
+
+	/* decrease the usage count, and allow suspend */
+	pm_runtime_put(ndev->dev.parent);
+}
+
+static unsigned long mcp25xxfd_oscout_recalc_rate(struct clk_hw *hw, unsigned long parent_rate)
+{
+	struct mcp25xxfd_priv *priv =
+		container_of(hw, struct mcp25xxfd_priv, oscout);
+
+	switch (priv->clkodiv) {
+	case MCP25XXFD_REG_OSC_CLKODIV_1:
+		return parent_rate;
+	case MCP25XXFD_REG_OSC_CLKODIV_2:
+		return parent_rate / 2;
+	case MCP25XXFD_REG_OSC_CLKODIV_4:
+		return parent_rate / 4;
+	case MCP25XXFD_REG_OSC_CLKODIV_10:
+		return parent_rate / 10;
+	}
+
+	return 0;
+}
+static const struct clk_ops mcp25xxfd_clk_ops = {
+	.prepare = mcp25xxfd_oscout_prepare,
+	.unprepare = mcp25xxfd_oscout_unprepare,
+	.recalc_rate = mcp25xxfd_oscout_recalc_rate,
+};
+
+static int mcp25xxfd_oscout_register_clk(struct mcp25xxfd_priv *priv)
+{
+	struct clk_init_data init = {};
+	struct device *dev = &priv->spi->dev;
+	const char *parent_name;
+	const char *clk_name = "canX_oscout";
+	int ret;
+
+	parent_name = __clk_get_name(priv->clk);
+	init.parent_names = &parent_name;
+	init.num_parents = 1;
+
+	device_property_read_string(dev, "clock-output-names", &clk_name);
+	init.name = clk_name;
+	init.ops = &mcp25xxfd_clk_ops;
+	init.flags = CLK_GET_RATE_NOCACHE;
+	priv->oscout.init = &init;
```

Comment:
Please move all static stuff into the C99 initializer above

Reply:
done.